### ISSUE TITLE
Feature: Allow testing

### ISSUE THREAD

**Issue by bebraw** [christianalfoni/webpack-bin#134]
It could be nice if it was possible to set up tests for bins. This kind of bins could easily become individual exercises (katas) to fork.

**Comment by christianalfoni** [comment]
So you basically want a separate section where you write a test for the bin? How you suggest what we test? Should we test everything from DOM, files etc.? Or should it be a special BIN with only one javascript file which has to return something?

**Comment by bebraw** [comment]
> So you basically want a separate section where you write a test for the bin?

Yeah. Some standard webpack setup would hook into it (think `npm test`) and you would then write your tests there. This feature probably takes some work to get right. Initially I would stash all test suites to that separate section and then execute them on change.

This could be taken far further by integrating code coverage and better test management. But the first version could be rough. I think it would be literally just another entry for webpack to execute. Now that I think of it, you might be able to hack it right now to some extent. I'm not so sure on the exact concept as there are many ways to approach this.

Perhaps looking from a kata perspective would be one way. See http://es6katas.org/ for an example. That's what I'm getting at here. I would love to be able to write small exercises with little tests to validate them.

**Comment by christianalfoni** [comment]
If you got time it would be great to try to "hack it" with current version. Then we can figure out minimum implementation to make that a bit easier. Then we can build from there. As you say, this could be a huge scope and hard to get right. So maybe just start with the minimal stuff?

If you check it out, please let me know how you did it and we can discuss what implementations we can do to make it better? :)

**Comment by bebraw** [comment]
Sure, I'll give it a go.

**Comment by bebraw** [comment]
I gave it a quick go. Mocha or tape won't work out of the box due to fs dependency. Here's a tape demo: http://www.webpackbin.com/EyFbZdZGZ .

The most likely scenario is that we'll need a little test harness of our own unless we can find one. It's enough to provide just enough structure that you can define test suites, individual tests, and asserts. It should also handle reporting through webpackbin so you get a good idea of what failed. That can be even integrated to the UI (coloring or some such).

**Comment by christianalfoni** [comment]
You know... we could create a `webpackbin-tests` npm module that you could add like any other module. That would give the API you need to write the tests. So could just use the `assert` library and add some methods to write the suits. Does that make sense?

**Comment by bebraw** [comment]
> You know... we could create a webpackbin-tests npm module that you could add like any other module. That would give the API you need to write the tests. So could just use the assert library and add some methods to write the suits. Does that make sense?

Yeah, that would be enough. You could write asserts using `chai` or something like that.

**Comment by christianalfoni** [comment]
Great, if you got time to look into such a module that would be great. Or I will look into it at some later point. Great discussion! :)

**Comment by bebraw** [comment]
Looks like we might not have to develop it after all. I found a standalone version of tape. Try http://www.webpackbin.com/EyFbZdZGZ . It outputs to browser but I assume it wouldn't require that much work to pipe the output to `bin.log` or elsewhere.

**Comment by christianalfoni** [comment]
Ah, thats awesome! Though isnt it good to just show it in preview window? As the tests you want to run are probably javascript related anyways?

But how do you do assignments test driven? You write the actual tests yourself as a student?

**Comment by bebraw** [comment]
> Ah, thats awesome! Though isnt it good to just show it in preview window? As the tests you want to run are probably javascript related anyways?

Yeah. The current solution is pretty good actually.

> But how do you do assignments test driven? You write the actual tests yourself as a student?

Assuming I'm teaching how to write React components, I would write the tests myself and let the student write components to match the specifications. But if you are teaching testing, then it makes sense to let the student write them (how to test tests? :) ).

**Comment by christianalfoni** [comment]
Hehehe, okay cool. So what you really want to do is prepare a plain React component in a file, then write a test which imports that file and runs the tests on the component?

**Comment by bebraw** [comment]
> Hehehe, okay cool. So what you really want to do is prepare a plain React component in a file, then write a test which imports that file and runs the tests on the component?

Correct. I wrote a little demo based on shallow testing, http://www.webpackbin.com/EyFbZdZGZ . That's quite neat already.

**Comment by christianalfoni** [comment]
Eyh, thats great! I suggest you give that a spin and come back with feedback on how it works and then we try to define how we can improve it? :)

**Comment by bebraw** [comment]
Yeah, I'll use this if/when I get more training opportunities.
